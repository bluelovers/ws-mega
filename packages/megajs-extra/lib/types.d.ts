/// <reference types="node" />
import { FileOptions, File, DownloadOptions } from 'megajs';
import { Readable } from "stream";
import { SymCryptoKey } from './util';
/**
 * Created by user on 2020/5/24.
 */
declare class FileChildren {
    constructor(options: FileOptions | string);
    static fromURL(options: FileOptions | string): File;
    static unpackAttributes(at: any): JSON;
    name: string;
    attributes: object;
    size: number;
    key: Buffer;
    timestamp: number;
    nodeId: string;
    downloadId: string[];
    directory: boolean;
    children: ReadonlyArray<IFileChildren>;
    parent?: ReadonlyArray<IFileChildren | IFile>;
    loadAttributes(cb?: (err: Error | null, file: IFileChildren) => any): Readable;
    download(options?: DownloadOptions, cb?: any): Readable;
    [SymCryptoKey]?: Buffer;
}
declare class File2 extends File {
    loadAttributes(cb?: (err: Error | null, file: IFile) => any): Readable;
    children: ReadonlyArray<IFileChildren>;
    [SymCryptoKey]?: Buffer;
}
export declare type IFileChildren = FileChildren;
export declare type IFile = File2;
export {};
