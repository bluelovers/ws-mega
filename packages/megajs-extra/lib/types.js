"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const megajs_1 = require("megajs");
const util_1 = require("./util");
//# sourceMappingURL=types.js.map