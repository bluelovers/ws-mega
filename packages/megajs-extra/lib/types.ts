import { FileOptions, File, DownloadOptions } from 'megajs';
import { Readable } from "stream";
import { SymCryptoKey } from './util';

/**
 * Created by user on 2020/5/24.
 */

declare class FileChildren
{
	constructor(options: FileOptions | string);
	static fromURL(options: FileOptions | string): File;
	static unpackAttributes(at: any): JSON;
	name: string;
	attributes: object;
	size: number;
	key: Buffer;
	timestamp: number;
	nodeId: string;
	downloadId: string[];
	directory: boolean;
	children: ReadonlyArray<IFileChildren>;
	parent?: ReadonlyArray<IFileChildren | IFile>;
	loadAttributes(cb?: (err: Error | null, file: IFileChildren) => any): Readable;
	download(options?: DownloadOptions, cb?: any): Readable;

	[SymCryptoKey]?: Buffer;
}

declare class File2 extends File
{
	/*
	constructor(options: FileOptions | string);
	static fromURL(options: FileOptions | string): File;
	static unpackAttributes(at: any): JSON;
	name: string;
	attributes: object;
	size: number;
	key: Buffer;
	timestamp: number;
	nodeId: string;
	downloadId: string;
	directory: boolean;
	children: ReadonlyArray<File>;
	download(options?: DownloadOptions, cb?: any): Readable;
	 */
	loadAttributes(cb?: (err: Error | null, file: IFile) => any): Readable;
	// @ts-ignore
	children: ReadonlyArray<IFileChildren>;

	[SymCryptoKey]?: Buffer;
}

export type IFileChildren = FileChildren;
export type IFile = File2;

export type IFileLike = IFile | IFileChildren | File
