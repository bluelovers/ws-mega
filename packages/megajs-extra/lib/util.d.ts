/**
 * Created by user on 2020/5/24.
 */
/// <reference types="node" />
import { IFile, IFileChildren } from './types';
export declare const SymCryptoKey: unique symbol;
export declare function e64(buffer: Buffer): string;
export declare function d64(key: string): Buffer;
export declare function megaKeyFromFile(file: IFile | IFileChildren): string;
export declare function megaLinkFromFile(file: IFile | IFileChildren, rootFile: IFile | IFileChildren, options?: {
    gateway?: string;
}): string;
export declare function megaFileList(file: IFile | IFileChildren, options?: {
    rootPath?: string;
    map?: Record<string, IFile | IFileChildren>;
    children?: boolean;
}): Record<string, IFile | IFileChildren>;
