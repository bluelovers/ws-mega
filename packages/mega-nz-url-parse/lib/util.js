"use strict";
/**
 * Created by user on 2020/5/24.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseLinkHash2 = exports.parseLinkHash = exports.parseSubPath = exports.validDownloadID = void 0;
function validDownloadID(file_hash) {
    return /^\w{8,}$/.test(file_hash);
}
exports.validDownloadID = validDownloadID;
function parseSubPath(paths) {
    if (!['file', 'folder'].includes(paths[0])) {
        throw TypeError(`Invalid URL: ${paths}`);
    }
    const downloadID = paths[1];
    if (!validDownloadID(downloadID)) {
        throw TypeError(`Invalid URL: not a valid file hash '${downloadID}'`);
    }
    const directory = paths[0] === 'folder';
    return {
        directory,
        downloadID,
    };
}
exports.parseSubPath = parseSubPath;
function parseLinkHash(hash) {
    if (hash[0] !== '#') {
        throw TypeError(`Invalid URL: invalid hash ${hash}`);
    }
    hash = hash.slice(1);
    const paths = hash.split('/');
    const key = paths[0];
    if (paths.length > 1) {
        return {
            ...parseSubPath(paths.slice(1)),
            key,
        };
    }
    else {
        return {
            key,
        };
    }
}
exports.parseLinkHash = parseLinkHash;
function parseLinkHash2(hash) {
    const split = hash.split('!');
    if (split[0] !== '#' && split[0] !== '#F') {
        throw TypeError(`Invalid URL: format not recognized ${hash}`);
    }
    if (split.length <= 1)
        throw TypeError(`Invalid URL: too few arguments ${hash}`);
    if (split.length >= (split[0] === '#' ? 4 : 5)) {
        throw TypeError(`Invalid URL: too many arguments ${hash}`);
    }
    const downloadID = split[1];
    const key = split[2];
    const directory = split[0] === '#F';
    const loadedFile = split[3];
    return {
        directory,
        downloadID,
        key,
        loadedFile,
    };
}
exports.parseLinkHash2 = parseLinkHash2;
//# sourceMappingURL=util.js.map