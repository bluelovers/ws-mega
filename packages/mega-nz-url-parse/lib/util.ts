/**
 * Created by user on 2020/5/24.
 */

export function validDownloadID(file_hash: string)
{
	return /^\w{8,}$/.test(file_hash)
}

export function parseSubPath(paths: string[])
{
	if (!['file', 'folder'].includes(paths[0]))
	{
		throw TypeError(`Invalid URL: ${paths}`)
	}

	const downloadID = paths[1];

	if (!validDownloadID(downloadID))
	{
		throw TypeError(`Invalid URL: not a valid file hash '${downloadID}'`)
	}

	const directory = paths[0] === 'folder';

	return {
		directory,
		downloadID,
	}
}

export function parseLinkHash(hash: string): {
	key: string;
	directory?: never;
	downloadID?: never;
} | {
	key: string;
	directory: boolean;
	downloadID: string;
}
{
	if (hash[0] !== '#')
	{
		throw TypeError(`Invalid URL: invalid hash ${hash}`)
	}

	hash = hash.slice(1);

	const paths = hash.split('/');

	const key = paths[0];

	if (paths.length > 1)
	{
		return {
			...parseSubPath(paths.slice(1)),
			key,
		}
	}
	else
	{
		return {
			key,
		}
	}
}

export function parseLinkHash2(hash: string)
{
	const split = hash.split('!')

	if (split[0] !== '#' && split[0] !== '#F') { throw TypeError(`Invalid URL: format not recognized ${hash}`) }
	if (split.length <= 1) throw TypeError(`Invalid URL: too few arguments ${hash}`)
	if (split.length >= (split[0] === '#' ? 4 : 5)) { throw TypeError(`Invalid URL: too many arguments ${hash}`) }

	const downloadID = split[1];
	const key = split[2];
	const directory = split[0] === '#F';
	const loadedFile = split[3];

	return {
		directory,
		downloadID,
		key,
		loadedFile,
	}
}
