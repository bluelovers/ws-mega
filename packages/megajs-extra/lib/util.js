"use strict";
/**
 * Created by user on 2020/5/24.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.megaFileList = exports.megaLinkFromFile = exports.megaKeyFromFile = exports.d64 = exports.e64 = exports.SymCryptoKey = void 0;
exports.SymCryptoKey = Symbol.for('root_key');
function e64(buffer) {
    return buffer.toString('base64')
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=/g, '');
}
exports.e64 = e64;
function d64(key) {
    return Buffer.from(key, 'base64');
}
exports.d64 = d64;
function megaKeyFromFile(file) {
    var _a;
    return e64((_a = file[exports.SymCryptoKey]) !== null && _a !== void 0 ? _a : file.key);
}
exports.megaKeyFromFile = megaKeyFromFile;
function megaLinkFromFile(file, rootFile, options) {
    var _a;
    const topkey = megaKeyFromFile(rootFile);
    let downloadId;
    if (typeof file.downloadId === 'string') {
        downloadId = file.downloadId;
    }
    else {
        // @ts-ignore
        downloadId = file.downloadId[file.downloadId.length - 1];
    }
    return new URL(`${(_a = options === null || options === void 0 ? void 0 : options.gateway) !== null && _a !== void 0 ? _a : `https://mega.nz/`}${(rootFile.directory
        ? `folder`
        : `file`)}/${rootFile.downloadId}#${topkey}/${(file.directory ? `folder` : `file`)}/${downloadId}`).href;
}
exports.megaLinkFromFile = megaLinkFromFile;
function megaFileList(file, options) {
    var _a;
    let map = (options === null || options === void 0 ? void 0 : options.map) || {};
    let rootPath = ((_a = options === null || options === void 0 ? void 0 : options.rootPath) !== null && _a !== void 0 ? _a : '') + file.name;
    if (!file.directory) {
        map[rootPath] = file;
    }
    else {
        if (options === null || options === void 0 ? void 0 : options.children) {
            rootPath += '/';
        }
        else {
            rootPath = '';
        }
        file
            .children
            .forEach((file) => {
            megaFileList(file, {
                rootPath,
                map,
                children: true,
            });
        });
    }
    return map;
}
exports.megaFileList = megaFileList;
//# sourceMappingURL=util.js.map