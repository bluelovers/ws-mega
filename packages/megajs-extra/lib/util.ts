/**
 * Created by user on 2020/5/24.
 */

import { File } from 'megajs';
import { IFile, IFileChildren } from './types';

export const SymCryptoKey = Symbol.for('root_key');

export function e64(buffer: Buffer)
{
	return buffer.toString('base64')
		.replace(/\+/g, '-')
		.replace(/\//g, '_')
		.replace(/=/g, '')
}

export function d64(key: string)
{
	return Buffer.from(key, 'base64')
}

export function megaKeyFromFile(file: IFile | IFileChildren)
{
	return e64(file[SymCryptoKey] ?? file.key)
}

export function megaLinkFromFile(file: IFile | IFileChildren, rootFile: IFile | IFileChildren, options?: {
	gateway?: string,
})
{
	const topkey = megaKeyFromFile(rootFile);

	let downloadId: string;

	if (typeof file.downloadId === 'string')
	{
		downloadId = file.downloadId;
	}
	else
	{
		// @ts-ignore
		downloadId = file.downloadId[file.downloadId.length - 1];
	}

	return new URL(`${options?.gateway ?? `https://mega.nz/`}${(rootFile.directory
		? `folder`
		: `file`)}/${rootFile.downloadId}#${topkey}/${(file.directory ? `folder` : `file`)}/${downloadId}`).href
}

export function megaFileList(file: IFile | IFileChildren, options?: {
	rootPath?: string,
	map?: Record<string, IFile | IFileChildren>,
	children?: boolean,
}): Record<string, IFile | IFileChildren>
{
	let map = options?.map || {};
	let rootPath = (options?.rootPath ?? '') + file.name;

	if (!file.directory)
	{
		map[rootPath] = file;
	}
	else
	{
		if (options?.children)
		{
			rootPath += '/';
		}
		else
		{
			rootPath = '';
		}

		file
			.children
			.forEach((file) =>
		{
			megaFileList(file, {
				rootPath,
				map,
				children: true,
			})
		})
		;
	}

	return map;
}
