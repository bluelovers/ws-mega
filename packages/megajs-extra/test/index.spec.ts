import { d64, e64 } from '../lib/util';

test(`crypto key`, () =>
{
	let key = `RNFlsWOf-QTOZJvsMmqLlA`;

	let actual = d64(key);
	let expected = key;

	expect(e64(actual)).toStrictEqual(expected);
	expect(actual).toMatchSnapshot();
	expect(e64(actual)).toMatchSnapshot();

});

