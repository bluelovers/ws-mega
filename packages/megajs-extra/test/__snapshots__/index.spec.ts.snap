// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crypto key 1`] = `
Object {
  "data": Array [
    68,
    209,
    101,
    177,
    99,
    159,
    249,
    4,
    206,
    100,
    155,
    236,
    50,
    106,
    139,
    148,
  ],
  "type": "Buffer",
}
`;

exports[`crypto key 2`] = `"RNFlsWOf-QTOZJvsMmqLlA"`;
